File: java/src/main/java/observer/Messenger.java
package observer;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class Messenger implements Subject {
    private final List<Observer> observerList = new ArrayList<>();

    @Override
    public void subscribe(Observer observer) {
        if (Objects.nonNull(observer)) {
            observerList.add(observer);
        }
    }

    @Override
    public void unsubscribe(Observer observer) {
       observerList.remove(observer);
    }

    @Override
    public void publish(String message) {
        for (var observer : observerList) {
            observer.update(message);
        }
    }
}

File: kotlin/src/main/kotlin/observer/Messenger.kt
package observer

class Messenger : Subject {
    private val observerList = ArrayList<Observer>()

    override fun subscribe(observer: Observer) {
        observerList.add(observer)
    }

    override fun unsubscribe(observer: Observer) {
        observerList.remove(observer)
    }

    override fun publish(message: String) {
        for (observer in observerList ) {
            observer.update(message)
        }
    }
}